platform-sdk/swirlds-benchmarks/src/jmh/java/com/swirlds/benchmark/ReconnectBench.java

```diff
@@ -71,62 +71,76 @@ public class ReconnectBench extends VirtualMapBaseBench {
-        final Path merkleDbPath = getTestDir().resolve("merkledb" + dbIndex++);
-        MerkleDb.setDefaultPath(merkleDbPath);
-    }
-
     /**
      * Builds a VirtualMap populator that is able to add/update, as well as remove nodes (when the value is null.)
      * Note that it doesn't support explicitly adding null values under a key.
      *
-     * @param map a VirtualMap instance
+     * @param mapRef a reference to a VirtualMap instance
      * @return a populator for the map
      * @param <K> key type
      * @param <V> value type
      */
     private static <K extends VirtualKey, V extends VirtualValue> BiConsumer<K, V> buildVMPopulator(
-            final VirtualMap<K, V> map) {
+            final AtomicReference<VirtualMap<K, V>> mapRef) {
         return (k, v) -> {
             if (v == null) {
-                map.remove(k);
+                mapRef.get().remove(k);
             } else {
-                map.put(k, v);
+                mapRef.get().put(k, v);
             }
         };
     }
 
-    @Setup(Level.Invocation)
-    public void setupInvocation() {
+    /** Generate a state and save it to disk once for the entire benchmark. */
+    @Setup
+    public void setupBenchmark() {
         beforeTest("reconnect");
+        updateMerkleDbPath();
 
-        teacherMap = createEmptyMap("teacher");
-        learnerMap = createEmptyMap("learner");
+        final AtomicReference<VirtualMap<BenchmarkKey, BenchmarkValue>> teacherRef =
+                new AtomicReference<>(createEmptyMap("teacher"));
+        final AtomicReference<VirtualMap<BenchmarkKey, BenchmarkValue>> learnerRef =
+                new AtomicReference<>(createEmptyMap("learner"));
 
         final Random random = new Random(randomSeed);
         new StateBuilder<>(BenchmarkKey::new, BenchmarkValue::new)
                 .buildState(
                         random,
-                        numRecords,
+                        (long) numRecords * (long) numFiles,
                         teacherAddProbability,
                         teacherRemoveProbability,
                         teacherModifyProbability,
-                        buildVMPopulator(teacherMap),
-                        buildVMPopulator(learnerMap));
+                        buildVMPopulator(teacherRef),
+                        buildVMPopulator(learnerRef),
+                        i -> {
+                            if (i % numRecords == 0) {
+                                System.err.printf("Copying files for i = %,d\n", i);
+                                teacherRef.set(copyMap(teacherRef.get()));
+                                learnerRef.set(copyMap(learnerRef.get()));
+                            }
+                        });
+
+        teacherRef.set(flushMap(teacherRef.get()));
+        learnerRef.set(flushMap(learnerRef.get()));
+
+        final List<VirtualMap<BenchmarkKey, BenchmarkValue>> mapCopies =
+                saveMaps(List.of(teacherRef.get(), learnerRef.get()));
```

Comment:
There is a benchmark config named `saveDataDirectory`. It makes sense to check it here

Reply:
While the parameter is undocumented, I believe its purpose if clear from reading

https://github.com/hashgraph/hedera-services/blob/develop/platform-sdk/swirlds-benchmarks/src/jmh/java/com/swirlds/benchmark/BaseBench.java#L164-L170

Basically it controls whether the benchmark would remove the data directory **after** the benchmark is complete. So it seems irrelevant to the saveMaps() calls - the ReconnectBench would still call this method regardless of the boolean value of that config parameter.

Reply:
If `saveDataDirectory` is true, a copy is saved after each test iteration, not just after the benchmark is complete

Reply:
I'm unable to find this particular logic. I see that flushMap() will saveMap() if it's true though.

However, the default value for the `saveDataDirectory` parameter is false as seen at:

https://github.com/hashgraph/hedera-services/blob/develop/platform-sdk/swirlds-benchmarks/src/jmh/java/com/swirlds/benchmark/config/BenchmarkConfig.java#L53

So this logic is never active by default. And the ReconnectBench doesn't use it explicitly, and doesn't recommend changing its default value.

So I still think that this parameter is completely irrelevant to the ReconnectBench. At least for now - until/unless we start using it in this particular benchmark.

Reply:
No, it is relevant. The logic is exactly the same as in virtual map benchmark: save the state, so the benchmark can be started from that state next time rather than from scratch

Reply:
I filed https://github.com/hashgraph/hedera-services/issues/12194 to address this at a later time when this feature becomes more relevant (which depends on the ability to generate much-much larger states.)